src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -341,6 +381,706 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
+	{
+		AttributeEquivalenceClass *attributeEquivalenceClass =
+			(AttributeEquivalenceClass *) lfirst(attributeEquivalenceClassCell);
+		ListCell *equivalenceMemberCell = NULL;
+		bool allPartitionKeys = true;
+
+		foreach(equivalenceMemberCell, attributeEquivalenceClass->equivalentAttributes)
+		{
+			AttributeEquivalenceClassMember *classMemeber =
+				(AttributeEquivalenceClassMember *) lfirst(equivalenceMemberCell);
+			Var *relationPartitionKey = PartitionKey(classMemeber->relationId);
+
+			/* we don't care about reference tables here */
+			if (!relationPartitionKey)
+			{
+				continue;
+			}
+
+			if (relationPartitionKey->varattno != classMemeber->varattno)
+			{
+				allPartitionKeys = false;
+				break;
+			}
+		}
+
+		if (!allPartitionKeys)
+		{
+			continue;
+		}
+
+		partitionKeyEquivalencClassList = lappend(partitionKeyEquivalencClassList,
+												  attributeEquivalenceClass);
+	}
+
+	return partitionKeyEquivalencClassList;
+}
+
+
+/*
+ * ReferenceRelationCount iterates over the relations and returns the reference table
+ * relation count.
+ */
+static uint32
+ReferenceRelationCount(RelationRestrictionContext *restrictionContext)
+{
+	ListCell *relationRestrictionCell = NULL;
+	uint32 referenceRelationCount = 0;
+
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction = lfirst(relationRestrictionCell);
+
+		if (PartitionMethod(relationRestriction->relationId) == DISTRIBUTE_BY_NONE)
+		{
+			referenceRelationCount++;
+		}
+	}
+
+	return referenceRelationCount;
+}
+
+
+/*
+ * GenerateAttributeEquivalencesForRelationRestrictions gets a relation restriction
+ * context and returns a list of AttributeEquivalenceClass.
+ *
+ * The algorithm followed can be summarized as below:
+ *
+ * - Per relation restriction
+ *     - Per plannerInfo's eq_class
+ *         - Create an AttributeEquivalenceClass
+ *         - Add all Vars that appear in the plannerInfo's
+ *           eq_class to the AttributeEquivalenceClass
+ *               - While doing that, consider LATERAL vars as well.
+ *                 See GetVarFromAssignedParam() for the details. Note
+ *                 that we're using parentPlannerInfo while adding the
+ *                 LATERAL vars given that we rely on that plannerInfo.
+ *
+ * Note that this function does not deal with whether the member of eq_classes
+ * are partition key or not. That's handled later in the planning.
+ */
+static List *
+GenerateAttributeEquivalencesForRelationRestrictions(RelationRestrictionContext
+													 *restrictionContext)
+{
+	List *attributeEquivalenceList = NIL;
+	ListCell *relationRestrictionCell = NULL;
+
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction = lfirst(relationRestrictionCell);
+		List *equivalenceClasses = relationRestriction->plannerInfo->eq_classes;
+		ListCell *equivilanceClassCell = NULL;
```

Comment:
>  > Doesn't this mean we're duplicating a lot of eq classes

> Yes, that's true. I can add some logic to prevent adding the same attributed eq class to the final output list. However, we'd be still processing the same eq_classes.

Yea, but we'd not duplicating things repeatedly.



    As I said, I think we should move to only collecting plannerInfos.

> But, wouldn't we still need to process each plannerInfo that is associated with a relation? (Or, do we skip the plannerInfos that are equal to each other?)

Yes, but I'm not seeing the problem?